### DISCUSSION TITLE
Is the ObjectNormalizer meant to be final ?

### DISCUSSION THREAD

**Discussion by andrei-dascalu** [symfony/symfony#50914] [Q&A] [answered]
I am wondering about this code in the serializer component

```php
/**
 * Converts between objects and arrays using the PropertyAccess component.
 *
 * @author Kévin Dunglas <dunglas@gmail.com>
 *
 * @final since Symfony 6.3
 */
class ObjectNormalizer extends AbstractObjectNormalizer
{
    protected $propertyAccessor;
....
}
```

It has the `@final` annotation, however the class is not actually final. This leads to analysis tools like psalm to interpret the class as final. Aside from the minor inconvenience, I am wondering whether something is missing here. The framework is at version 6.3.1 yet this class is not final, just annotated as such.

Will it become final later on? Was it meant to become final but then the decision was backtracked on? Is the annotation meaningful in itself? Basically, should the annotation be removed or should the final keyword be added? Should I open a bug on this?

Cheers

**Comment by MatTheCat** [comment, marked as answer]
Symfony follows semantic versioning, which means they can only break BC (backwards compatibility) when releasing major versions (see https://symfony.com/doc/current/contributing/code/bc.html). Making a class `final` breaks BC because userland classes which extends it would suddenly break.

`ObjectNormalizer` was annotated `@final` by #49953 which landed in 6.3.0, so you have until Symfony 7 to update your code. Until then, analysis tools will interpret the class as `final` thanks to the annotation, which lets you know which parts of your code need to be changed.

**Reply by andrei-dascalu** [reply]
@MatTheCat makes sense, thanks. It was a bit misleading because the text says it's final since 6.3 which kindof obscures the idea that it will become final actually in 7.

**Reply by stof** [reply]
it is `@final` since 6.3.
